lib/diagnostics_channel.js

```diff
@@ -0,0 +1,292 @@
+        return resolve({
+          value: message,
+          done: false
+        });
+      }
+
+      this.events.push(message);
+    };
+
+    channel.subscribe(this.subscriber);
+  }
+
+  [Symbol.asyncIterator]() {
+    return this;
+  }
+
+  return() {
+    const data = { done: true };
+    this.done = true;
+
+    this.channel.unsubscribe(this.subscriber);
+
+    for (let i = 0; i < this.waiting.length; i++) {
+      const resolve = this.waiting[i];
+      resolve(data);
+    }
+
+    return PromiseResolve(data);
+  }
+
+  next() {
+    const event = this.events.shift();
+    if (event) {
+      return PromiseResolve({
+        value: event,
+        done: false
+      });
+    }
+
+    if (this.done) {
+      return PromiseResolve({
+        done: true
+      });
+    }
+
+    return new Promise((resolve) => {
+      this.waiting.push(resolve);
+    });
+  }
+}
+
+// TODO(qard): should there be a C++ channel interface?
+class ActiveChannel {
+  [Symbol.asyncIterator]() {
+    return new AsyncIterableChannel(this);
+  }
+
+  subscribe(subscription) {
+    ArrayPrototypePush(this._subscribers, subscription);
+  }
+
+  unsubscribe(subscription) {
+    const index = this._subscribers.indexOf(subscription);
+    if (index >= 0) {
+      ArrayPrototypeSplice(this._subscribers, index, 1);
+
+      // When there are no more active subscribers, restore to fast prototype.
+      if (!this._subscribers.length) {
+        // eslint-disable-next-line no-use-before-define
+        ObjectSetPrototypeOf(this, Channel.prototype);
+      }
+    }
+  }
+
+  shouldPublish() {
+    return true;
+  }
+
+  publish(data) {
+    for (var i = 0; i < this._subscribers.length; i++) {
+      try {
+        const onMessage = this._subscribers[i];
+        onMessage(data, this.name);
+      } catch (err) {
+        process.nextTick(() => {
+          triggerUncaughtException(err, false);
+        });
+      }
+    }
+  }
+}
+
+class Channel {
+  constructor(name) {
+    this._subscribers = undefined;
+    this.name = name;
+  }
+
+  [Symbol.asyncIterator]() {
+    return new AsyncIterableChannel(this);
+  }
+
+  subscribe(subscription) {
+    ObjectSetPrototypeOf(this, ActiveChannel.prototype);
+    this._subscribers = [];
+    this.subscribe(subscription);
+  }
+
+  shouldPublish() {
+    return false;
+  }
+
+  publish() {}
+}
+
+class DiagnosticsChannel {
+  constructor() {
+    this._channels = ObjectCreate(null);
+    this._subscribers = [];
+  }
+
+  channel(name) {
+    let channel;
+    const ref = this._channels[name];
+    if (ref) channel = ref.get();
+    if (channel) return channel;
+
+    if (typeof name !== 'string' && typeof name !== 'symbol') {
+      throw new ERR_INVALID_ARG_TYPE('channel', ['string', 'symbol'], name);
+    }
+
+    channel = new Channel(name);
+
+    // New channels should check existing subscribers
+    for (let i = 0; i < this._subscribers.length; i++) {
+      const [matcher, onMessage] = this._subscribers[i];
```

Comment:
Avoid mutated Array.prototype[Symbol.iterator], might also be nicer to use an object with named fields

```suggestion
      const {0: matcher, 1: onMessage} = this._subscribers[i];
```